File: app/src/main/java/me/yxp/qfun/javaplugin/api/PluginMethod.java
package me.yxp.qfun.javaplugin.api;

import android.app.Activity;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

import me.yxp.qfun.javaplugin.loader.LoadJarHelper;
import me.yxp.qfun.javaplugin.loader.PluginCompiler;
import me.yxp.qfun.javaplugin.loader.PluginInfo;
import me.yxp.qfun.utils.error.PluginError;
import me.yxp.qfun.utils.json.JsonConfigUtils;
import me.yxp.qfun.utils.qq.CookieTool;
import me.yxp.qfun.utils.qq.FriendTool;
import me.yxp.qfun.utils.qq.MsgTool;
import me.yxp.qfun.utils.qq.QQCurrentEnv;
import me.yxp.qfun.utils.qq.ToastUtils;
import me.yxp.qfun.utils.qq.TroopTool;
import me.yxp.qfun.utils.thread.SyncUtils;

public class PluginMethod {
    private final PluginCompiler mPluginCompiler;
    private final PluginInfo mPluginInfo;
    private final String mConfigPath;

    public PluginMethod(PluginCompiler pluginCompiler) {
        mPluginCompiler = pluginCompiler;
        mPluginInfo = pluginCompiler.pluginInfo;
        mConfigPath = mPluginInfo.pluginPath + "config/";
    }

    // 插件管理方法
    public void loadJar(String jarPath) {
        try {
            mPluginCompiler.fixClassLoader.addClassLoader(LoadJarHelper.loadJar(jarPath));
        } catch (Exception e) {
            PluginError.callError(e, mPluginInfo);
        }
    }

    public void loadJava(String path) {
        try {
            mPluginCompiler.interpreter.source(path);
        } catch (Exception e) {
            PluginError.callError(e, mPluginInfo);
        }
    }

    // UI相关方法
    public void Toast(Object message) {
        ToastUtils.Toast(message.toString());
    }

    public void QQToast(int icon, Object message) {
        ToastUtils.QQToast(icon, message.toString());
    }

    public Activity getNowActivity() {
        return QQCurrentEnv.getActivity();
    }

    // 菜单项管理
    public void addItem(String name, String callback) {
        mPluginCompiler.addItem(name, callback);
    }

    public void addMenuItem(String name, String callback) {
        mPluginCompiler.addMenuItem("[QFun]," + mPluginInfo.pluginId + "," + name, callback);
    }

    public void addMenuItem(String name, String callback, int[] msgTypes) {
        String types = Arrays.stream(msgTypes).mapToObj(String::valueOf).collect(Collectors.joining(","));
        mPluginCompiler.addMenuItem("[QFun]," + mPluginInfo.pluginId + "," + name + "," + types, callback);
    }

    // 好友相关方法
    public List<HashMap<String, Object>> getAllFriend() {
        return executeWithErrorHandling(FriendTool::getAllFriend, null);
    }

    public boolean isFriend(String uin) {
        return executeWithErrorHandling(() -> FriendTool.isFriend(uin), false);
    }

    public String getUidFromUin(String uin) {
        return executeWithErrorHandling(() -> FriendTool.getUidFromUin(uin), "");
    }

    public String getUinFromUid(String uid) {
        return executeWithErrorHandling(() -> FriendTool.getUidFromUin(uid), "");
    }

    public void sendZan(String uin, int num) {
        executeWithErrorHandling(() -> FriendTool.sendZan(uin, num));
    }

    // 群组管理方法
    public void shutUp(String troopUin, String uin, long time) {
        executeWithErrorHandling(() -> TroopTool.shutUp(troopUin, uin, time));
    }

    public void shutUpAll(String troopUin, boolean bool) {
        executeWithErrorHandling(() -> TroopTool.shutUpAll(troopUin, bool));
    }

    public boolean isShutUp(String troopUin) {
        return executeWithErrorHandling(() -> TroopTool.isShutUp(troopUin), false);
    }

    public void setGroupAdmin(String troopUin, String uin, boolean bool) {
        executeWithErrorHandling(() -> TroopTool.setGroupAdmin(troopUin, uin, bool));
    }

    public void kickGroup(String troopUin, String uin, boolean bool) {
        executeWithErrorHandling(() -> TroopTool.kickGroup(troopUin, uin, bool));
    }

    public void setGroupMemberTitle(String troopUin, String uin, String title) {
        executeWithErrorHandling(() -> TroopTool.setGroupMemberTitle(troopUin, uin, title));
    }

    public List<HashMap<String, Object>> getGroupList() {
        return executeWithErrorHandling(TroopTool::getGroupList, null);
    }

    public List<HashMap<String, Object>> getGroupMemberList(String troopUin) {
        return executeWithErrorHandling(() -> TroopTool.getGroupMemberList(troopUin), null);
    }

    public List<HashMap<String, Object>> getProhibitList(String troopUin) {
        return executeWithErrorHandling(() -> TroopTool.getProhibitList(troopUin), null);
    }

    public Object getTroopInfo(String troopUin) {
        return executeWithErrorHandling(() -> TroopTool.getTroopInfo(troopUin), null);
    }

    public Object getMemberInfo(String troopUin, String uin) {
        return executeWithErrorHandling(() -> TroopTool.getMemberInfo(troopUin, uin), null);
    }

    public String getMemberName(String troopUin, String uin) {
        return executeWithErrorHandling(() -> TroopTool.getMemberName(troopUin, uin), "");
    }

    public void changeMemberName(String troopUin, String uin, String name) {
        executeWithErrorHandling(() -> TroopTool.changeMemberName(troopUin, uin, name));
    }

    // 消息管理方法
    public void recallMsg(int type, String peer, ArrayList<Long> list) {
        executeWithErrorHandling(() -> MsgTool.recallMsg(type, peer, list));
    }

    public void recallMsg(Object contact, ArrayList<Long> list) {
        executeWithErrorHandling(() -> MsgTool.recallMsg(contact, list));
    }

    public void sendMsg(String peerUin, String msg, int type) {
        executeWithErrorHandling(() -> MsgTool.sendMsg(peerUin, msg, type));
    }

    public void sendMsg(Object contact, String msg) {
        executeWithErrorHandling(() -> MsgTool.sendMsg(contact, msg));
    }

    public void sendPic(String peerUin, String path, int type) {
        executeWithErrorHandling(() -> MsgTool.sendPic(peerUin, path, type));
    }

    public void sendPic(Object contact, String path) {
        executeWithErrorHandling(() -> MsgTool.sendPic(contact, path));
    }

    public void sendPtt(String peerUin, String path, int type) {
        executeWithErrorHandling(() -> MsgTool.sendPtt(peerUin, path, type));
    }

    public void sendPtt(Object contact, String path) {
        executeWithErrorHandling(() -> MsgTool.sendPtt(contact, path));
    }

    public void sendCard(String peerUin, String data, int type) {
        executeWithErrorHandling(() -> MsgTool.sendCard(peerUin, data, type));
    }

    public void sendCard(Object contact, String data) {
        executeWithErrorHandling(() -> MsgTool.sendCard(contact, data));
    }

    public void sendVideo(String peerUin, String path, int type) {
        executeWithErrorHandling(() -> MsgTool.sendVideo(peerUin, path, type));
    }

    public void sendVideo(Object contact, String path) {
        executeWithErrorHandling(() -> MsgTool.sendVideo(contact, path));
    }

    public void sendFile(String peerUin, String path, int type) {
        executeWithErrorHandling(() -> MsgTool.sendFile(peerUin, path, type));
    }

    public void sendFile(Object contact, String path) {
        executeWithErrorHandling(() -> MsgTool.sendFile(contact, path));
    }

    public void sendReplyMsg(String peerUin, long replyMsgId, String msg, int type) {
        executeWithErrorHandling(() -> MsgTool.sendReplyMsg(peerUin, replyMsgId, msg, type));
    }

    public void sendReplyMsg(Object contact, long replyMsgId, String msg) {
        executeWithErrorHandling(() -> MsgTool.sendReplyMsg(contact, replyMsgId, msg));
    }

    // Cookie相关方法
    public String getRealSkey() {
        return executeWithErrorHandling(CookieTool::getRealSkey, null);
    }

    public String getSkey() {
        return executeWithErrorHandling(CookieTool::getSkey, null);
    }

    public String getPskey(String url) {
        return executeWithErrorHandling(() -> CookieTool.getPskey(url), null);
    }

    public String getGTK(String url) {
        return executeWithErrorHandling(() -> CookieTool.getGTK(url), null);
    }

    public String getStweb() {
        return executeWithErrorHandling(CookieTool::getStweb, null);
    }

    public String getPt4Token(String url) {
        return executeWithErrorHandling(() -> CookieTool.getPt4Token(url), null);
    }

    public long getBkn(String key) {
        return executeWithErrorHandling(() -> CookieTool.getBkn(key), 0L);
    }

    public String getFriendRKey() {
        return executeWithErrorHandling(CookieTool::getFriendRKey, "");
    }

    public String getGroupRKey() {
        return executeWithErrorHandling(CookieTool::getGroupRKey, "");
    }

    // 配置管理方法
    public void putString(String configName, String key, String value) {
        executeWithErrorHandling(() -> JsonConfigUtils.putString(mConfigPath, configName, key, value));
    }

    public void putInt(String configName, String key, int value) {
        executeWithErrorHandling(() -> JsonConfigUtils.putInt(mConfigPath, configName, key, value));
    }

    public void putBoolean(String configName, String key, boolean value) {
        executeWithErrorHandling(() -> JsonConfigUtils.putBoolean(mConfigPath, configName, key, value));
    }

    public void putLong(String configName, String key, long value) {
        executeWithErrorHandling(() -> JsonConfigUtils.putLong(mConfigPath, configName, key, value));
    }

    public String getString(String configName, String key, String defaultValue) {
        return executeWithErrorHandling(() ->
                JsonConfigUtils.getString(mConfigPath, configName, key, defaultValue), defaultValue);
    }

    public int getInt(String configName, String key, int defaultValue) {
        return executeWithErrorHandling(() ->
                JsonConfigUtils.getInt(mConfigPath, configName, key, defaultValue), defaultValue);
    }

    public boolean getBoolean(String configName, String key, boolean defaultValue) {
        return executeWithErrorHandling(() ->
                JsonConfigUtils.getBoolean(mConfigPath, configName, key, defaultValue), defaultValue);
    }

    public long getLong(String configName, String key, long defaultValue) {
        return executeWithErrorHandling(() ->
                JsonConfigUtils.getLong(mConfigPath, configName, key, defaultValue), defaultValue);
    }

    // 统一的错误处理方法
    private void executeWithErrorHandling(SyncUtils.MyRunnable action) {
        try {
            action.run();
        } catch (Throwable throwable) {
            PluginError.callError(throwable, mPluginInfo);
        }
    }

    private <T> T executeWithErrorHandling(SupplierWithException<T> supplier, T defaultValue) {
        try {
            return supplier.get();
        } catch (Exception e) {
            PluginError.callError(e, mPluginInfo);
            return defaultValue;
        }
    }

    @FunctionalInterface
    private interface SupplierWithException<T> {
        T get() throws Exception;
    }
}

File: app/src/main/java/me/yxp/qfun/plugin/api/PluginMethod.kt
@file:Suppress("unused")

package me.yxp.qfun.plugin.api

import android.app.Activity
import bsh.classpath.BshLoaderManager
import com.tencent.mobileqq.data.troop.TroopInfo
import com.tencent.qqnt.kernelpublic.nativeinterface.Contact
import me.yxp.qfun.common.ModuleScope
import me.yxp.qfun.plugin.bean.ForbidInfo
import me.yxp.qfun.plugin.bean.FriendInfo
import me.yxp.qfun.plugin.bean.GroupInfo
import me.yxp.qfun.plugin.bean.MemberInfo
import me.yxp.qfun.plugin.loader.JarLoader
import me.yxp.qfun.plugin.loader.PluginCompiler
import me.yxp.qfun.utils.io.FileUtils
import me.yxp.qfun.utils.json.JsonConfigUtils
import me.yxp.qfun.utils.log.PluginError
import me.yxp.qfun.utils.qq.CookieTool
import me.yxp.qfun.utils.qq.FriendTool
import me.yxp.qfun.utils.qq.MsgTool
import me.yxp.qfun.utils.qq.QQCurrentEnv
import me.yxp.qfun.utils.qq.Toasts
import me.yxp.qfun.utils.qq.TroopTool
import me.yxp.qfun.utils.reflect.ClassUtils
import java.io.File

class PluginMethod(private val compiler: PluginCompiler) {

    private val configPath = "${compiler.info.dirPath}/config/"

    @JvmOverloads
    fun log(fileName: String = "log.txt", msg: String) {
        FileUtils.writeText(File(compiler.info.dirPath, fileName), "$msg\n", true)
    }

    fun getNowActivity(): Activity? = QQCurrentEnv.activity

    fun toast(msg: Any?) {
        Toasts.toast("$msg")
    }

    fun qqToast(icon: Int, msg: Any?) {
        Toasts.qqToast(icon, "$msg")
    }

    fun addItem(name: String, callback: String) {
        compiler.menuItems[name] = callback
    }

    @JvmOverloads
    fun addMenuItem(name: String, callback: String, msgTypes: IntArray = intArrayOf()) {
        val types = msgTypes.joinToString(",")
        compiler.msgMenuItem.add("[QFun],${compiler.info.id},$name,$callback,$types")
    }

    fun loadJava(path: String) {
        runWithErrorHandle { compiler.interpreter.source(path) }
    }

    fun loadJar(path: String) {
        runWithErrorHandle {
            compiler.loader.addClassLoader(JarLoader.loadJar(path))
        }
    }

    fun loadDex(path: String) {
        runWithErrorHandle {
            compiler.loader.addClassLoader(
                BshLoaderManager.getDexLoader(path, ClassUtils.hostClassLoader)
            )
        }
    }

    fun getAllFriend(): List<FriendInfo>? = runWithErrorHandle(null, FriendTool::getAllFriend)

    fun isFriend(uin: String): Boolean = runWithErrorHandle(false) { FriendTool.isFriend(uin) }

    fun getUidFromUin(uin: String): String =
        runWithErrorHandle("") { FriendTool.getUidFromUin(uin) }

    fun getUinFromUid(uid: String): String =
        runWithErrorHandle("") { FriendTool.getUinFromUid(uid) }

    fun sendZan(uin: String, num: Int) {
        runWithErrorHandle { FriendTool.sendZan(uin, num) }
    }

    fun clockIn(troopUin: String) {
        runWithErrorHandle { TroopTool.clockIn(troopUin) }
    }

    fun getGroupList(): List<GroupInfo>? = runWithErrorHandle(null, TroopTool::getGroupList)

    fun getGroupInfo(troopUin: String): TroopInfo? =
        runWithErrorHandle(null) { TroopTool.getGroupInfo(troopUin) }

    fun getGroupMemberList(troopUin: String): List<MemberInfo>? =
        runWithErrorHandle(null) { TroopTool.getGroupMemberList(troopUin) }

    fun getProhibitList(troopUin: String): List<ForbidInfo>? =
        runWithErrorHandle(null) { TroopTool.getProhibitList(troopUin) }

    fun getMemberInfo(troopUin: String, uin: String): MemberInfo? =
        runWithErrorHandle(null) { TroopTool.getMemberInfo(troopUin, uin) }

    fun isShutUp(troopUin: String): Boolean =
        runWithErrorHandle(false) { TroopTool.isShutUp(troopUin) }

    fun shutUpAll(troopUin: String, enable: Boolean) {
        runWithErrorHandle { TroopTool.shutUpAll(troopUin, enable) }
    }

    fun shutUp(troopUin: String, uin: String, time: Long) {
        runWithErrorHandle { TroopTool.shutUp(troopUin, uin, time) }
    }

    fun setGroupAdmin(troopUin: String, uin: String, enable: Boolean) {
        runWithErrorHandle { TroopTool.setGroupAdmin(troopUin, uin, enable) }
    }

    fun setGroupMemberTitle(troopUin: String, uin: String, title: String) {
        ModuleScope.launchMain {
            runWithErrorHandle { TroopTool.setGroupMemberTitle(troopUin, uin, title) }
        }
    }

    fun changeMemberName(troopUin: String, uin: String, name: String) {
        runWithErrorHandle { TroopTool.changeMemberName(troopUin, uin, name) }
    }

    fun kickGroup(troopUin: String, uin: String, block: Boolean) {
        runWithErrorHandle { TroopTool.kickGroup(troopUin, uin, block) }
    }

    fun sendPai(toUin: String, peerUin: String, chatType: Int) {
        runWithErrorHandle { MsgTool.sendPai(toUin, peerUin, chatType) }
    }

    fun recallMsg(chatType: Int, peerUin: String, msgId: Long) {
        runWithErrorHandle { MsgTool.recallMsg(chatType, peerUin, msgId) }
    }

    fun recallMsg(contact: Contact, msgId: Long) {
        runWithErrorHandle { MsgTool.recallMsg(contact, msgId) }
    }

    fun sendMsg(peerUin: String, msg: String, chatType: Int) {
        runWithErrorHandle { MsgTool.sendMsg(peerUin, msg, chatType) }
    }

    fun sendMsg(contact: Contact, msg: String) {
        runWithErrorHandle { MsgTool.sendMsg(contact, msg) }
    }

    fun sendPic(peerUin: String, path: String, chatType: Int) {
        runWithErrorHandle { MsgTool.sendPic(peerUin, path, chatType) }
    }

    fun sendPic(contact: Contact, path: String) {
        runWithErrorHandle { MsgTool.sendPic(contact, path) }
    }

    fun sendPtt(peerUin: String, path: String, chatType: Int) {
        runWithErrorHandle { MsgTool.sendPtt(peerUin, path, chatType) }
    }

    fun sendPtt(contact: Contact, path: String) {
        runWithErrorHandle { MsgTool.sendPtt(contact, path) }
    }

    fun sendCard(peerUin: String, data: String, chatType: Int) {
        runWithErrorHandle { MsgTool.sendCard(peerUin, data, chatType) }
    }

    fun sendCard(contact: Contact, data: String) {
        runWithErrorHandle { MsgTool.sendCard(contact, data) }
    }

    fun sendVideo(peerUin: String, path: String, chatType: Int) {
        runWithErrorHandle { MsgTool.sendVideo(peerUin, path, chatType) }
    }

    fun sendVideo(contact: Contact, path: String) {
        runWithErrorHandle { MsgTool.sendVideo(contact, path) }
    }

    fun sendFile(peerUin: String, path: String, chatType: Int) {
        runWithErrorHandle { MsgTool.sendFile(peerUin, path, chatType) }
    }

    fun sendFile(contact: Contact, path: String) {
        runWithErrorHandle { MsgTool.sendFile(contact, path) }
    }

    fun sendReplyMsg(peerUin: String, replyMsgId: Long, msg: String, chatType: Int) {
        runWithErrorHandle { MsgTool.sendReplyMsg(peerUin, replyMsgId, msg, chatType) }
    }

    fun sendReplyMsg(contact: Contact, replyMsgId: Long, msg: String) {
        runWithErrorHandle { MsgTool.sendReplyMsg(contact, replyMsgId, msg) }
    }

    fun putString(configName: String, key: String, value: String) {
        runWithErrorHandle { JsonConfigUtils.putString(configPath, configName, key, value) }
    }

    fun putInt(configName: String, key: String, value: Int) {
        runWithErrorHandle { JsonConfigUtils.putInt(configPath, configName, key, value) }
    }

    fun putBoolean(configName: String, key: String, value: Boolean) {
        runWithErrorHandle { JsonConfigUtils.putBoolean(configPath, configName, key, value) }
    }

    fun putLong(configName: String, key: String, value: Long) {
        runWithErrorHandle { JsonConfigUtils.putLong(configPath, configName, key, value) }
    }

    fun getString(configName: String, key: String, defaultValue: String): String =
        runWithErrorHandle(defaultValue) {
            JsonConfigUtils.getString(
                configPath, configName, key, defaultValue
            )
        }

    fun getInt(configName: String, key: String, defaultValue: Int): Int =
        runWithErrorHandle(defaultValue) {
            JsonConfigUtils.getInt(
                configPath, configName, key, defaultValue
            )
        }

    fun getBoolean(configName: String, key: String, defaultValue: Boolean): Boolean =
        runWithErrorHandle(defaultValue) {
            JsonConfigUtils.getBoolean(
                configPath, configName, key, defaultValue
            )
        }

    fun getLong(configName: String, key: String, defaultValue: Long): Long =
        runWithErrorHandle(defaultValue) {
            JsonConfigUtils.getLong(
                configPath, configName, key, defaultValue
            )
        }

    fun getRealSkey(): String? = runWithErrorHandle("", CookieTool::getRealSkey)

    fun getSkey(): String? = runWithErrorHandle("", CookieTool::getSkey)

    fun getStweb(): String? = runWithErrorHandle("", CookieTool::getStweb)

    fun getPt4Token(url: String): String? = runWithErrorHandle("") { CookieTool.getPt4Token(url) }

    fun getGTK(url: String): String = runWithErrorHandle("") { CookieTool.getGTK(url) }

    fun getBkn(key: String): Long = runWithErrorHandle(0L) { CookieTool.getBkn(key) }

    fun getGroupRKey(): String = runWithErrorHandle("", CookieTool::getGroupRKey)

    fun getFriendRKey(): String = runWithErrorHandle("", CookieTool::getFriendRKey)

    fun getPskey(url: String): String? = runWithErrorHandle("") { CookieTool.getPskey(url) }

    private inline fun <T> runWithErrorHandle(default: T, function: () -> T): T = runCatching {
        function()
    }.onFailure {
        PluginError.callError(it, compiler.info)
    }.getOrElse { default }

    private inline fun runWithErrorHandle(function: () -> Unit) = try {
        function()
    } catch (t: Throwable) {
        PluginError.callError(t, compiler.info)
    }


}